Sources/WasmKitGDBHandler/DebuggerMemoryCache.swift

```diff
@@ -0,0 +1,108 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if WasmDebuggingSupport
+
+    import NIOCore
+    import WasmKit
+
+    package struct DebuggerMemoryCache: ~Copyable {
+        package static let executableCodeOffset = UInt64(0x4000_0000_0000_0000)
+
+        private let allocator: ByteBufferAllocator
+        private let stackOffsetInProtocolSpace: UInt64
+
+        /// WebAssembly binary loaded into memory for execution
+        /// and for disassembly by the debugger.
+        private let wasmBinary: ByteBuffer
+
+        /// Cached packed stack frames that were previously requested.
+        private var stackFrames: ByteBuffer
+
+        /// Mapping from frame index to a base address of a frame in `stackFrames` buffer with
+        /// its corresponding layout.
+        private var stackFrameLayouts = [Int: (frameBase: Int, offsets: DebuggerStackFrame.Layout)]()
+
+        package init(allocator: ByteBufferAllocator, wasmBinary: ByteBuffer) {
+            self.allocator = allocator
+
+            var stackOffset = Int(Self.executableCodeOffset) + wasmBinary.readableBytes
+            // Untyped raw Wasm values in VM's stack are stored as `UInt64`.
+            stackOffset.roundUpToAlignment(for: UInt64.self)
+
+            self.stackOffsetInProtocolSpace = UInt64(stackOffset)
+            self.wasmBinary = wasmBinary
+            self.stackFrames = allocator.buffer(capacity: 0)
```

Comment:
I think it's better to simply use Optional here instead of allocating 0 so that we don't need to remember about `malloc(0)` semantics 😅